### ISSUE TITLE
mns-api请求次数计费的项目

### ISSUE THREAD

**Issue by carl-leopard** [gogap/ali_mns#5]
mns中有计费组成部分
请求次数: MNS对每个计量周期内的 API 请求次数进行计量，以百万次为单位，精确到小数点后三位。
但是在项目， 希望能将MNS中拉取的数据当做影子设备，用来减少影子设备相关的http请求；同时也减少api的请求次数。 那请求的频率是多少合适呢？

**Comment by xujinzheng** [comment]
可以使用批量获取接口，这个要减少到多少次，得看你们的实际业务场景吧。

**Comment by carl-leopard** [comment]
使用mns作为设备的实时状态， 这个影响有多大呢？我想将mns中的设备状态消息存储， 作为设备的最新状态（例：空调：制冷 27℃）， 后面再对设备进行操控的时候， 直接以本地最新的数据作为当前设备状态，而不去影子设备查询。 这样的错误时多高呢？

**Comment by xujinzheng** [comment]
我理解的是，你是想把你的每个设备，用MNS上报给服务器端是吧？然后服务端获取MNS里的数据作为初始数据，然后后续的业务操作？

**Comment by carl-leopard** [comment]
不是。 app端操控设备的流程是先向阿里云请求设备状态， 然后将更改对应的属性值， 再写入到影子设备。现在我想通过MNS获取设备的全部消息（上报消息和状态变更消息）， 然后app端在操控设备的时候， 不用向阿里云请求设备状态， 而是直接以我本地存储的设备为准， 改变对应的属性值， 最后写入影子设备。 这样就减少了影子设备信息的请求，同时也达到了信息落地的需要， 从而降低费用的目的。

**Comment by xujinzheng** [comment]
再理解一下，你是通过APP获取最新设备状态，然后通过APP控制空调设备，同时想将设备的状态同步到服务器端的影子设备上，以便下次APP启动的时候获取最新的设备状态？

**Comment by carl-leopard** [comment]
是的， app端要实时的显示设备的当前状态， 但是我想app显示的信息以我本地的数据为准， 而不用去阿里云的IoT套件获取

**Comment by xujinzheng** [comment]
个人思路如下(请求示意)：

设备数据上报：
Device -> MNS -> Server [持久化到MySql]

APP状态控制：
APP -> MNS -> Server [持久化到MySql](这里存的每一次变化)

设备获取App的最新控制指令
Device -> API(自己写一个API，获取到最终的设备状态, 当然如果业务需要，也可以是把每个步骤都下发) -> Device - > API( 再次上报设备当前状态)

App获取最新控制指令

App -> API (获取当前设备状态) -> App


最关键的还是数据同步的一致性问题，上报数据用MNS，因为MNS有过期时间，以及最少投递等。

但是需要注意的是，MNS不保证投递顺序，所以 服务端在收到 MNS 的数据后，要做一下排序的操作。

**Comment by carl-leopard** [comment]
是的， 主要是他不保证FIFO， 很坑。 
APP状态控制：
APP -> MNS -> Server 持久化到MySql， 我的理解是： APP -> 私有云(我们自己加入的， 作权限控制) ->IoT套件（http服务）-> 影子设备 -> 同步更新设备状态 -> 设备状态上报
这个流程没问题吧？

**Comment by xujinzheng** [comment]
差不多是这个流程吧，根据你们自己的业务场景调整吧。

**Comment by carl-leopard** [comment]
在项目中有一个QPS的监控， 个人不是很清楚QPS在这个项目中的作用。 还望解答一下。还有这个DefaultQPSLimit      int32 = 2000。 是干什么用的？

**Comment by xujinzheng** [comment]
哦，这个是控制客户端请求每秒请求频率的，当然了，MNS服务本身在服务端也时候有做限制的。

**Comment by carl-leopard** [comment]
懂了，谢谢😃

**Comment by carl-leopard** [comment]
请问一下， 主题队列的部分有吗？

**Comment by carl-leopard** [comment]
请问一下， 我目前使用的是阿里云的 IoT 套件 + MNS 消息队列 的组合进行物联网消息的传输与日志采集， 但是现在 IoT 套件设置的规则是 产品与服务端订阅（ mns 队列）一一对应的关系。 怎样做， 我才能将所有的产品的日志全部转发到 同一个 MNS 消息队列呢？你之前有过这样的经验吗？

**Comment by xujinzheng** [comment]
我没用过IoT，帮不到你，你可以尝试按照他们的案例做一下

**Comment by carl-leopard** [comment]
好的， 谢谢。 建议你们如果需要使用 IoT 的时候， 可以对比看一下腾讯的。 阿里云官方推荐 Iot + MNS 是个大坑。 很多细节他们不开放。
